include/zephyr/net/net_l2_connectivity.h

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Public API for managing L2 layers that have separated post-admin-up connect
+ *	  or associate phases
+ */
+
+#ifndef ZEPHYR_INCLUDE_NET_L2_CONNECTIVITY_H_
+#define ZEPHYR_INCLUDE_NET_L2_CONNECTIVITY_H_
+
+#include <zephyr/device.h>
+#include <zephyr/net/net_if.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+
+/**
+ * @brief Network Layer 2 connectivity management abstraction layer
+ * @defgroup net_l2_connectivity Network L2 Connectivity Abstraction Layer
+ * @ingroup networking
+ * @{
+ */
+
+/**
+ * @brief Network L2 connectivity API structure
+ *
+ * Used to provide an interface to connect parameters and procedures
+ */
+struct net_l2_connectivity_api {
+	int (*connect)(struct net_if *iface);
+	int (*disconnect)(struct net_if *iface);
+
+	/* Note to reviewers:
+	 * Technically, as long as each iface provides a means by which to set
+	 * association/connection parameters, these get/set_opt funcs are not needed.
+	 *
+	 * However, it can be argued that they should still exist, since they establish a
+	 * standardized pattern for doing that.
+	 *
+	 * The question, then, is: Why limit these to specifically connectivity?
+	 * Should these apply to the whole iface, and be distributed to implementations
+	 * (where they are present) on not only the connectivity_api, but also the l2_api
+	 * and if_api as well?
+	 *
+	 * For now, I have implemented them just on the connectivity_api, but I wonder if they
+	 * should perhaps be more generalized.
+	 */
+	int (*set_opt)(int optname, const void *optval, size_t optlen);
+	int (*get_opt)(int optname, const void *optval, size_t *optlen);
+};
+
+/**
+ * @brief Network interface connectivity structure
+ *
+ * Binds L2 connectivity API and settings to an iface / network device
+ */
+struct net_if_conn {
+	/* The network device the connectivity struct is bound to */
+	struct net_if_dev *if_dev;
```

Comment:
Orignally I though it'd be implemented the other way around (`net_if_dev` keeping a pointer to `net_if_conn`) but this could also work. Maybe it's even nicer, as there is no RAM penalty (pointer) for interfaces that do not use L2 connectivity.

Reply:
That is what I had been thinking too (and what I had implemented until recently), but the issue I encountered is that it becomes impossible to have a separate macro for registering the connectivity API since C doesn't allow structs to be modified at build time, and neither does it offer any mechanism to create mixin initializers.

So if we went the other route (net_if_dev keeping a pointer to net_if_conn), then we could not have the separated NET_DEVICE_DEFINE_L2_CONNECTIVITY macro, and net_if_conn would instead have to be an additional argument to NET_DEVICE_INIT and its ilk.

Going via the section lists is pretty much the only work-around to this issue I could find that operates at compile-time, but it is possible I'm just missing some other technique that could accomplish it! Do you know of any?

I'm fine going with whichever solution folks here think would be cleanest / most canonical. This was just my attempt to avoid further complicating the NET_DEVICE_INIT argument list

Reply:
I'm personally fine with the current approach. I'd be more concerned if the connectivity API was heavily used due to additional overhead for finding the corresponding `net_if_conn` structure, but it's not the case here.